jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java

```diff
@@ -0,0 +1,371 @@
+                    heavy.get(v).add(e);
+                } else {
+                    light.get(v).add(e);
+                }
+            }
+        });
+    }
+
+    /**
+     * Performs shortest paths computation.
+     *
+     * @param source the source vertex
+     */
+    private void computeShortestPaths(V source) {
+        relax(source, null, 0.0);
+
+        int firstNonEmptyBucket = 0;
+        while (firstNonEmptyBucket != -1) {
+            List<V> removed = new ArrayList<>();
+            List<V> bucketElements = bucketElements(firstNonEmptyBucket);
+            while (!bucketElements.isEmpty()) {
+                removed.addAll(bucketElements);
+                clearBucket(firstNonEmptyBucket);
+
+                findAndRelaxRequests(bucketElements, light);
+                bucketElements = bucketElements(firstNonEmptyBucket);
+            }
+            findAndRelaxRequests(removed, heavy);
+            firstNonEmptyBucket = firstNonEmptyBucket();
+        }
+    }
+
+    /**
+     * For each vertex v in {@code vertices} relaxes all edges emanating from v
+     * that are present in {@code edgesKind#get(v)}.
+     * Elements in {@code vertices} are processed in parallel.
+     *
+     * @param vertices  vertices
+     * @param edgesKind vertex to edges map
+     */
+    private void findAndRelaxRequests(List<V> vertices, Map<V, Set<E>> edgesKind) {
+        vertices.parallelStream().forEach(v -> {
+            for (E e : edgesKind.get(v)) {
+                relax(Graphs.getOppositeVertex(graph, e, v), e,
+                        verticesDataMap.get(v).get().getSecond() + graph.getEdgeWeight(e));
+            }
+        });
+    }
+
+    /**
+     * Performs relaxation in parallel-safe fashion. Vertex data {@code v} is
+     * considered updated once either {@code distance} is greater than or equal
+     * to tentative distance in {@link #verticesDataMap} or
+     * {@link AtomicReference#compareAndSet(Object, Object)} returned {@code true}.
+     *
+     * @param v        vertex
+     * @param e        edge to predecessor
+     * @param distance distance
+     */
+    private void relax(V v, E e, double distance) {
+        boolean updated = false;
+        AtomicReference<Triple<Integer, Double, E>> dataReference = verticesDataMap.get(v);
+        Triple<Integer, Double, E> updatedData = Triple.of(bucketIndex(distance), distance, e);
+        while (!updated) {
+            Triple<Integer, Double, E> oldData = dataReference.get();
+            if (distance < oldData.getSecond()) {
+                updated = dataReference.compareAndSet(oldData, updatedData);
+            } else {
+                updated = true;
+            }
+        }
+    }
+
+    /**
+     * Calculates num of buckets in the bucket structure.
+     *
+     * @return num of buckets
+     */
+    private int numOfBuckets() {
+        return getMaxEdgeWeight().map(maxWeight -> (int) (Math.ceil(maxWeight / delta) + 1)).orElse(1);
+    }
```

Comment:
To my mind, it is not necessary to changed  `getMaxEdgeWeight()` result type to `OptionalDouble` , because it will result in more sophisticated code here, since `OptionalDouble`  does not have `map()` method.

Reply:
Okay, that sounds perfectly reasonable to me! I consider `map()` to be an important method, so I'm happy for `Optional<Double>` to be returned here.